### PR TITLE
Proposal Remove the weird `inspect` in ASR pipeline and make WhisperEncoder just nice to use.

### PR DESCRIPTION
**Pull request by Narsil**
# What does this PR do?

It seems that accepting `attention_mask` is kind of an invariant of our
models. For Seq2Seq ASR models, we had a special comment on how it
actually was important to send it.

`inspecting` seems pretty brittle way to handle this case.
My suggestion is to simply add it as an kwarg that and just ignoring
it with the docstring explaining why it's ignored.

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: src/transformers/models/whisper/modeling_whisper.py [hunk 1]
```diff
@@ -631,12 +632,13 @@ def forward(
                 `input_features`, the [`WhisperFeatureExtractor`] should be used for extracting the mel features,
                 padding and conversion into a tensor of type `torch.FloatTensor`. See
                 [`~WhisperFeatureExtractor.__call__`]
+            attention_mask (`torch.Tensor`)`, *optional*):
+                Not actually used, pad_tokens are ignored by default. The argument is preserved for compatibility.
```
Comment (LysandreJik) [inline src/transformers/models/whisper/modeling_whisper.py:636]: As a user I wouldn't understand that :smile:

Comment (LysandreJik) [reply src/transformers/models/whisper/modeling_whisper.py:636]: (neither do I understand it as a reviewer not involved in the initial contribution)

File: src/transformers/pipelines/automatic_speech_recognition.py [hunk 2]
```diff
@@ -305,15 +300,15 @@ def _forward(self, model_inputs):
...
-                tokens = self.model.generate(inputs)
+            # we need to pass `processed.get("attention_mask")` here since audio encoder
+            # attention mask  length is different from expected text decoder `encoder_attention_mask` length
+            # `generate` magic to create the mask automatically won't work, we basically need to help
+            # it here.
+            attention_mask = model_inputs.pop("attention_mask", None)
+            tokens = self.model.generate(
+                encoder_outputs=encoder(inputs, attention_mask=attention_mask),
+                attention_mask=attention_mask,
+            )
```
Comment (LysandreJik) [inline src/transformers/pipelines/automatic_speech_recognition.py:311]: Very clean!

File: src/transformers/models/whisper/modeling_whisper.py (hunk 1 above)
Comment (Narsil) [reply src/transformers/models/whisper/modeling_whisper.py:636]: We could also just not document it. 

But `attention_mask` is used throughout the library to tell models to skip some `input_ids`.
I think not allowing it is not really respecting the invariant here.

Comment (ArthurZucker) [reply src/transformers/models/whisper/modeling_whisper.py:636]: We could deprecate the usage directly : 
```suggestion
            attention_mask (`torch.Tensor`)`, *optional*):
                Whisper does not support masking of the `input_features`, this argument is preserved for compatibility, but it is not used. By default the silence in the input log mel spectrogram are ignored. 
```

Comment (ArthurZucker) [reply src/transformers/models/whisper/modeling_whisper.py:636]: But I agree that having the argument makes a cleaner pipeline call. I am happy either way

**Review by ArthurZucker — APPROVED** [review]
LGTM, would still ping @sgugger to get his opinion on the matter 😉

File: src/transformers/models/whisper/modeling_whisper.py [hunk 3]
```diff
@@ -631,12 +632,13 @@ def forward(
                 `input_features`, the [`WhisperFeatureExtractor`] should be used for extracting the mel features,
                 padding and conversion into a tensor of type `torch.FloatTensor`. See
                 [`~WhisperFeatureExtractor.__call__`]
+            attention_mask (`torch.Tensor`)`, *optional*):
+                Whisper does not support masking of the `input_features`, this argument is preserved for compatibility, but it is not used. By default the silence in the input log mel spectrogram are ignored.
```
Comment (sgugger) [inline src/transformers/models/whisper/modeling_whisper.py:636]: `make style` will fix this veeeeeery long line ;-)

**Review by sgugger — APPROVED** [review]
Very clean!

File: src/transformers/models/whisper/modeling_whisper.py (hunk 3 above)
Comment (Narsil) [reply src/transformers/models/whisper/modeling_whisper.py]: It was a good suggestion I accepted from GH suggestion from @ArthurZucker .

:)

**Pull request merged**